File: J15/main.py
# Test grossier des bench avec le même algorithme
#
# Pour la partie 2 en python: ~8.8 secondes
#
# Pour la partie 2, même processeur, conditions similaires, même algorithme en JavaScript: ~286.4 secondes

f = open('input.txt', 'r')
input = list(map(int, f.read().split(',')))
f.close()

def resolve(n):
    fq = dict()
    prev = input[0]
    current = 0
    for i in range(1, n):
        if i >= len(input):
            if prev in fq:
                current = i - fq[prev]
            else:
                current = 0
        else:
            current = input[i]
        fq[prev] = i
        prev = current
    return current

print(resolve(2020))
print(resolve(30000000))

File: J15/main.js
const getInput = require("../injectInput")

/*
Utilisation de Map à la place d'un objet pour la HashTable ->
de ~286s à ~3.9s
*/

const resolve = (input, n) => {
    input = input[0].split(',').map(Number)
    let fq = new Map(), prev = input[0], current
    for(let i=1; i<n; i++) {
        if(!input[i]) current = fq.get(prev) ? i - fq.get(prev) : 0
        else current = input[i]
        fq.set(prev, i)
        prev = current
    }
    return current
}

(async () => {
    const input = await getInput(String)
    console.log("Part1: " +resolve(input, 2020))
    console.time('TimeP2')
    console.log("Part2: " +resolve(input, 30000000))
    console.timeEnd('TimeP2')
})()
